nix-rust/src/lib.rs

```diff
@@ -0,0 +1,32 @@
+mod error;
+mod foreign;
+mod tarfile;
+
+pub use error::Error;
+
+pub struct CBox<T> {
+    pub ptr: *mut libc::c_void,
+    phantom: std::marker::PhantomData<T>,
+}
+
+impl<T> CBox<T> {
+    fn new(t: T) -> Self {
+        unsafe {
+            let size = std::mem::size_of::<T>();
+            let ptr = libc::malloc(size);
+            *(ptr as *mut T) = t; // FIXME: probably UB
```

Comment:
This is indeed UB, as you're dereferencing an arbitrary value (which could be illegal for this `T`); this could cause double-frees, arbitrary UAFs, segfaults, etc.